hugegraph-core/src/main/java/org/apache/hugegraph/traversal/optimize/TraversalUtil.java

```diff
@@ -105,12 +107,71 @@ public static HugeGraph getGraph(Step<?, ?> step) {
     }
 
     public static HugeGraph tryGetGraph(Step<?, ?> step) {
-        Graph graph = step.getTraversal().getGraph().get();
-        if (graph instanceof HugeGraph) {
-            return (HugeGraph) graph;
+        Optional<Graph> graph = step.getTraversal()
+                                    .getGraph()
+                                    .filter(g -> {
+                                        return !(g instanceof EmptyGraph);
+                                    });
+        if (!graph.isPresent()) {
+            TraversalParent parent = step.getTraversal().getParent();
+            if (parent instanceof Traversal) {
+                Optional<Graph> parentGraph = ((Traversal<?, ?>) parent)
+                                              .asAdmin()
+                                              .getGraph()
+                                              .filter(g -> {
+                                                 return !(g instanceof EmptyGraph);
+                                              });
+                if (parentGraph.isPresent()) {
+                    step.getTraversal().setGraph(parentGraph.get());
+                    return (HugeGraph) parentGraph.get();
+                }
+            }
+
+            return null;
         }
-        assert graph == null || graph instanceof EmptyGraph;
-        return null;
+
+        assert graph.get() instanceof HugeGraph;
+        return (HugeGraph) graph.get();
+    }
+
+    public static void trySetGraph(Step<?, ?> step, HugeGraph graph) {
```

Comment:
seems don't need to set graph anymore, please refer to https://github.com/apache/tinkerpop/pull/1699

Reply:
ok, I have add TODO
        // TODO: remove these EmptyGraph judgments when upgrading tinkerpop to a fixed version